src/controller/CommissioningWindowOpener.cpp

```diff
@@ -0,0 +1,350 @@
+#include <protocols/secure_channel/PASESession.h>
+#include <setup_payload/ManualSetupPayloadGenerator.h>
+#include <setup_payload/QRCodeSetupPayloadGenerator.h>
+
+using namespace chip::app::Clusters;
+using namespace chip::System::Clock;
+
+namespace {
+// TODO: What should the timed invoke timeout here be?
+constexpr uint16_t kTimedInvokeTimeoutMs = 10000;
+} // anonymous namespace
+
+namespace chip {
+namespace Controller {
+
+CHIP_ERROR CommissioningWindowOpener::OpenBasicCommissioningWindow(NodeId deviceId, Seconds16 timeout,
+                                                                   Callback::Callback<OnOpenBasicCommissioningWindow> * callback)
+{
+    VerifyOrReturnError(mNextStep == Step::kAcceptCommissioningStart, CHIP_ERROR_INCORRECT_STATE);
+    mSetupPayload = SetupPayload();
+
+    // Basic commissioning does not use the setup payload.
+
+    mCommissioningWindowOption        = CommissioningWindowOption::kOriginalSetupCode;
+    mBasicCommissioningWindowCallback = callback;
+    mCommissioningWindowCallback      = nullptr;
+    mNodeId                           = deviceId;
+    mCommissioningWindowTimeout       = timeout;
+
+    mNextStep = Step::kOpenCommissioningWindow;
+    return mController->GetConnectedDevice(mNodeId, &mDeviceConnected, &mDeviceConnectionFailure);
+}
+
+CHIP_ERROR CommissioningWindowOpener::OpenCommissioningWindow(NodeId deviceId, Seconds16 timeout, uint32_t iteration,
+                                                              uint16_t discriminator, Optional<uint32_t> setupPIN,
+                                                              Callback::Callback<OnOpenCommissioningWindow> * callback,
+                                                              SetupPayload & payload, bool readVIDPIDAttributes)
+{
+    VerifyOrReturnError(mNextStep == Step::kAcceptCommissioningStart, CHIP_ERROR_INCORRECT_STATE);
+
+    mSetupPayload = SetupPayload();
+
+    if (setupPIN.HasValue())
+    {
+        mCommissioningWindowOption = CommissioningWindowOption::kTokenWithProvidedPIN;
+        mSetupPayload.setUpPINCode = setupPIN.Value();
+    }
+    else
+    {
+        mCommissioningWindowOption = CommissioningWindowOption::kTokenWithRandomPIN;
+    }
+
+    mSetupPayload.version               = 0;
+    mSetupPayload.discriminator         = discriminator;
+    mSetupPayload.rendezvousInformation = RendezvousInformationFlags(RendezvousInformationFlag::kOnNetwork);
+
+    mCommissioningWindowCallback      = callback;
+    mBasicCommissioningWindowCallback = nullptr;
+    mNodeId                           = deviceId;
+    mCommissioningWindowTimeout       = timeout;
+    mCommissioningWindowIteration     = iteration;
+
+    bool randomSetupPIN = !setupPIN.HasValue();
+    ReturnErrorOnFailure(PASESession::GeneratePASEVerifier(mVerifier, mCommissioningWindowIteration, GetSPAKE2Salt(),
+                                                           randomSetupPIN, mSetupPayload.setUpPINCode));
```

Comment:
We should no longer do this. It's already handled by Linux that the verifier is always computed